### PR TITLE
Change onnxruntime::make_unique to std::make_unique

### PR DESCRIPTION
**Pull request by snnn**
**Description**: 

1. Change onnxruntime::make_unique to std::make_unique
2. Add "-std=c++14" to ROCM EP's build flags.

I didn't delete "include/onnxruntime/core/common/make_unique.h" in case someone is adding new code to master and using it. I will delete it in a subsequent PR.

**Motivation and Context**
- Why is this change required? What problem does it solve?

As we have dropped support for GCC 4.x, now we have better C++14 support and we can directly use std::make_unique.

- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: cmake/onnxruntime_providers.cmake [hunk 1]
```diff
@@ -1030,8 +1030,7 @@ if (onnxruntime_USE_ROCM)
   # Generate GPU code for GFX9 Generation
   list(APPEND HIP_CLANG_FLAGS --amdgpu-target=gfx906 --amdgpu-target=gfx908)
 
-  hip_add_library(onnxruntime_providers_rocm ${onnxruntime_providers_rocm_src})
-
+  hip_add_library(onnxruntime_providers_rocm ${onnxruntime_providers_rocm_src} HIPCC_OPTIONS -std=c++14 NVCC_OPTIONS -std=c++14)
```
Comment (weixingzhang) [inline cmake/onnxruntime_providers.cmake:1033]: Could you add the line below to line#1019 rather than change here? 

list(APPEND HIP_CXX_FLAGS -std=c++14)

Comment (snnn) [reply cmake/onnxruntime_providers.cmake]: Updated.

Comment (weixingzhang) [reply cmake/onnxruntime_providers.cmake]: Thanks!

**Review by weixingzhang — APPROVED** [review]

**Comment by edgchen1** [comment]
cool, glad we can use std::make_unique now.
can we also update this line?
```
docs/Coding_Conventions_and_Standards.md:29:* If there is a legitimate need to allocate objects on the heap, prefer using onnxruntime::make_unique(). References for the reasoning:
```
also, it seems like breaking someone's pending change by removing make_unique.h now or later isn't too different, so why not do it in one PR?

**Comment by snnn** [comment]
Doc updated.

> Why not do it in one PR?
This PR is very big, I expect it will take quite a while to check in.  Whenever I made a change, I need to ask you help me review this huge change again. I feel it would be easier if I put the deletion in another PR.

File: include/onnxruntime/core/framework/execution_provider.h [hunk 2]
```diff
@@ -6,7 +6,7 @@
 #ifndef SHARED_PROVIDER
 #include <unordered_map>
 #include <unordered_set>
-
+#include <memory>
```
Comment (edgchen1) [inline include/onnxruntime/core/framework/execution_provider.h:9]: nit: this could be moved outside of the ifdefs

**Review by edgchen1 — APPROVED** [review]

**Review by thiagocrepaldi — APPROVED** [review]
How about deleting `onnxruntime::make_unique()` to prevent accidental use in future PRs?

**Review by BowenBao — APPROVED** [review]

**Pull request merged**